Allow building static CQiree

ausbin opened the pull request:
With these changes, the compilation flow I need in the Qwerty compiler works. That is, I can create a [`qiree_sys` crate][1] that [builds a static QIR-EE][2] and provides Rust bindings, call QIR-EE from Rust code, and link the whole thing into a Python extension module (a self-contained shared library).

I think this PR is easiest reviewed commit-by-commit. I tried to leave an explanation of each change in each commit.

#### Testing

The unit tests pass and `bin/qir-qsim ../examples/bv11010.ll` runs successfully. (Note that I built only with `-DQIREE_USE_QSIM=ON -DQIREE_USE_XACC=OFF`, though.)

I can also [call `qiree_create()` and `qiree_destroy()` from inside the Qwerty runtime][3] without any obvious explosions.

[1]: https://github.com/gt-tinker/qwerty/tree/70b2075ce699155408c9b228cc03d4ab265f1f9c/qiree_sys
[2]: https://github.com/gt-tinker/qwerty/blob/70b2075ce699155408c9b228cc03d4ab265f1f9c/qiree_sys/build.rs#L52-L67
[3]: https://github.com/gt-tinker/qwerty/blob/70b2075ce699155408c9b228cc03d4ab265f1f9c/qwerty_ast_to_mlir/src/run/backend/qiree.rs#L9-L12

ausbin committed:
Allow building static cqiree

Passing `MODULE` to `add_library` seems to force a shared library to be
built even if `BUILD_SHARED_LIBS=OFF`.

ausbin committed:
Install CQiree headers

Without this, the C headers are not installed.

ausbin committed:
Remove a C++ include inside a C header

`#include <cstdint>` does not compile on a C compiler.

ausbin committed:
Mention that LLVM 21 works

ausbin committed:
Add `#include` to get CQiree tests to build

Without this, the C++ compiler on my GNU/Linux system cannot find
`std::sort()`.

sethrj commented on src/cqiree/CMakeLists.txt:
```diff
@@ -12,7 +12,7 @@ if(QIREE_USE_QSIM)
 endif()
 
-qiree_add_library(cqiree MODULE
+qiree_add_library(cqiree
```
We *might* want to make this configurable. There are a few subtle differences between a module and a shared library, and in the current setup, if you make the default libraries static, the cqiree test should fail because a static library (rather than dynamic) will be created.

ausbin replied:
Thanks for calling that out, you're right. I tried to add an option `QIREE_SHARED_LIBS` to control this, although I'm not sure if that's what you had in mind.

Locally the tests pass with either shared or static libs now (before I was only testing shared libs, but I didn't make that clear above). I also added static builds to the build matrix.

sethrj commented on src/cqiree/CQiree.h:
Your fix tells me of another issue here: these headers shouldn't be inside the extern"C"! Could you mvoe them above it? Thanks!

sethrj reviewed (commented):
Thanks for the fixes! (If you move those to a separate PR I can approve and merge them immediately.) The one-line change removing `MODULE` may have consequences that are untested: I'm actually a little surprised it works as is in all configurations, which might just mean we need a little more testing (e.g. a runner that builds static libs).

ausbin committed:
Move `#include`s outside `extern "C"`

Per PR comment.

ausbin committed:
Handle JSON in .clang-format

Otherwise, when modifying CMakePresets.json and running `git
clang-format`, I get:

    Configuration file(s) do(es) not support Json

Also update the pre-commit-hooks configuration to allow the second YAML
document that this commit adds in .clang-format.

ausbin committed:
Make static build controlled by config option

ausbin committed:
Add static build to build matrix

sethrj commented:
Please avoid force pushing! 🙂

ausbin commented:
Why?

sethrj commented on CMakeLists.txt:
```diff
@@ -38,6 +38,7 @@ option(QIREE_BUILD_TESTS "Build QIR-EE unit tests" ON)
 option(QIREE_BUILD_EXAMPLES "Build QIR-EE examples" OFF)
 option(QIREE_USE_QSIM "Download and build Google qsim backend" OFF)
 option(QIREE_USE_XACC "Build XACC interface" ON)
+option(QIREE_SHARED_LIBS "Build shared libraries, not static libraries" ON)
```
Instead of defining a new option, use the built-in CMake `BUILD_SHARED_LIBS`. Or alternatively, define a `QIREE_CQIREE_MODULE` boolean.

sethrj commented:
@ausbin It interferes with the review process. When you force push:
1. It becomes difficult/impossible to tell changes you've made since the last review.
2. The automatic record of whether your previous changes passed or failed the CI are lost.
3. The reviewer clicking on a notification on the update for the PR gets a page that says "the commit range can't be found" rather than a page showing the changes you pushed.
4. You are noncompliant with policy in the CONTRIBUTING guidelines (which also documents most of these reasons)

ausbin commented:
Thanks for sharing your perspective. It sounds like we have different styles of using git — agree to disagree.

In this case, I was constructing a readable commit history while also fighting with CI. When you have time, please feel free to check out the four new commits added today. I did not modify previous commits.

sethrj commented:
@ausbin I use force push all the time and am a huge supporter of rebasing, but *only before a review has started*. I personally do prefer readable commit histories, but after a review starts the review process needs to become part of the history.

Because we use the "squash" method, the changes and merges and so forth done during review are silently flattened into a single clean commit which is the only history most people see. The iterations made during review are saved in this pull request's commit history which is also very useful.

You can disagree with style and policies, but part of working in an open source collaboration is agreeing to follow the development practices required by the project.

sethrj commented on .github/workflows/build-fast.yml:
```diff
@@ -10,18 +10,30 @@ concurrency:
+            compiler: gcc
+            version: 12
+            llvm: 14
+            shared: OFF
```
Use boolean literals true/false for this

sethrj commented on .github/workflows/build-fast.yml:
```diff
@@ -10,18 +10,30 @@ concurrency:
 jobs:
   linux:
-    name: ${{matrix.runner}}-${{matrix.compiler}}-${{matrix.version}}-llvm${{matrix.llvm}}
+    name: ${{matrix.runner}}-${{matrix.compiler}}-${{matrix.version}}-${{ matrix.shared == 'ON' && 'shared' || 'static' }}-llvm${{matrix.llvm}}
```
With a bool, this will simplify:
```suggestion
    name: ${{matrix.runner}}-${{matrix.compiler}}-${{matrix.version}}-${{ matrix.shared && 'shared' || 'static' }}-llvm${{matrix.llvm}}
```

sethrj commented on .github/workflows/build-fast.yml:
```diff
@@ -10,18 +10,30 @@ concurrency:
+            compiler: clang
+            version: 15
+            llvm: 15
+            shared: ON
```
I don't think the linker changes when we use the different toolchains, so I believe it's necessary only to have a single static build (one new CI job).

sethrj commented on src/cqiree/CMakeLists.txt:
```diff
@@ -12,7 +12,15 @@ if(QIREE_USE_QSIM)
+else()
+    set(_LIB_TYPE)
+endif()
+
```
Format with two spaces indentation

sethrj commented on src/cqiree/CMakeLists.txt:
```diff
@@ -28,10 +36,10 @@ target_link_libraries(cqiree
+  DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/cqiree"
   COMPONENT development
-  FILES_MATCHING REGEX ".*\\.hh$"
+  FILES_MATCHING REGEX ".*\\.hh?$"
```
My bad for blindly copy-pasting this originally!

Should  we be installing the `.hh` files if this is meant to be a C-only/module interface? Perhaps we should just install the C header to avoid confusion.

sethrj commented on src/qiree_config.h.in:
```diff
@@ -12,5 +12,6 @@
 #cmakedefine01 QIREE_DEBUG
 #cmakedefine01 QIREE_USE_QSIM
 #cmakedefine01 QIREE_USE_XACC
+#cmakedefine01 QIREE_SHARED_LIBS
```
I don't think the configuration should include the library linkage details since linking doesn't affect the source/build. Hoewever, since the test needs this information, it's appropriate to add this to the qiree_test_config.h.in.

sethrj commented on test/CMakeLists.txt:
```diff
@@ -69,7 +69,13 @@ qiree_add_test(qiree ResultDistribution)
+else()
+    # Statically link tester with cqiree
+    target_link_libraries(cqiree_CQireeTest cqiree)
+endif()
```
Indentation

sethrj commented on CMakeLists.txt:
```diff
@@ -77,10 +78,10 @@ qiree_set_default(CMAKE_CXX_EXTENSIONS OFF)
 # Inform installed binaries of external library rpaths
 qiree_set_default(CMAKE_INSTALL_RPATH_USE_LINK_PATH ON)
-if(BUILD_SHARED_LIBS)
+if(QIREE_SHARED_LIBS)
```
I think just use the standard variables.

sethrj commented on src/cqiree/CMakeLists.txt:
```diff
@@ -12,7 +12,15 @@ if(QIREE_USE_QSIM)
+    set(_LIB_TYPE)
+endif()
+
+qiree_add_library(cqiree ${_LIB_TYPE}
```
Another alternative to this is to use an object library that allows us to build a dlopen-compatible module *and* a linkable (shared or static) library... @miniskar thoughts?

ausbin replied:
Let me know if you want me to attempt this

ausbin committed:
Address PR comments
